src/main/java/org/jsoup/helper/W3CDom.java

```diff
@@ -348,6 +387,7 @@ protected static class W3CBuilder implements NodeVisitor {
         public W3CBuilder(Document doc) {
             this.doc = doc;
             namespacesStack.push(new HashMap<>());
+            namespacesStack.peek().put("", "http://www.w3.org/1999/xhtml"); // TODO document
```

Comment:
I feel that we should only be setting this default namespace if the input document is HTML. If it's XML, then the HTML -> XML compat note doesn't apply, and we shouldn't be applying it.

The impl could be something like
```
if (inDoc.parser().getTreeBuilder() instanceof HtmlTreeBuilder)
```
Would need to move this to the appropriate `convert()` method. Or, add that flag as user data to the W3C Document.

I'm not sure if checking the treebuilder is the best way to see if it's HTML vs XML. The output syntax is how we normally check that, but that doesn't tell us particularly if the input was HTML, and people may be setting the syntax to XML before calling these methods. So am leaning towards checking treebuilder.

Reply:
> I feel that we should only be setting this default namespace if the input document is HTML.

That makes sense. I don't know enough about your library's API to know when it thinks it's parsing XML. I had assumed it only parsed things considered to be non-XML HTML. (Otherwise I would think you could just use an off-the-shelf XML parser because the input would not be "soup".)

> So am leaning towards checking treebuilder.

You know best here. All I care about is when I parse an HTML document, the HTML namespace gets imputed as per the space. If the tree builder is always set to `HtmlTreeBuilder` in those cases, that works. Whatever you want to do.

> Would need to move this to the appropriate `convert()` method. Or, add that flag as user data to the W3C Document.

That sounds like a lot of shuffling. I note that the `W3CBuilder` does seem to get passed the `org.jsoup.nodes.Element` being converted via the W3C document user data context property. In normal cases that element would know its owner `org.jsoup.nodes.Document` would it not? (If not, can we assume this is HTML input? I would imagine it to be a rare case anyway for the jsoup element not to have an owner document.)

So what about this inside the `W3CBuilder` constructor (where I made the change already)?

```java
final org.jsoup.nodes.Document inDoc = contextElement.ownerDocument();
if (inDoc == null || inDoc.parser().getTreeBuilder() instanceof HtmlTreeBuilder) {
  namespacesStack.peek().put("", "http://www.w3.org/1999/xhtml"); // TODO document
}
```

(I see that I had forgotten to document the actual namespace imputation line. I'll add that once I get an OK from you on how to proceed.)

Reply:
@jhy could you respond here and let me know which way to go, so that I can finish this PR and get it accepted?

Reply:
> That sounds like a lot of shuffling. I note that the `W3CBuilder` does seem to get passed the `org.jsoup.nodes.Element` being converted via the W3C document user data context property. In normal cases that element would know its owner `org.jsoup.nodes.Document` would it not? (If not, can we assume this is HTML input? I would imagine it to be a rare case anyway for the jsoup element not to have an owner document.)

OK, using the contextElement is a good idea. And yes let's go with checking if the builder was HtmlTreeBuilder. That will work for all cases in the core library (vs anyone's own extensions, which is kind of up them).

I would not assume that an Element without an ownerDocument is HTML though, so we should not add the namespace if it is null. There will be no ownerDoc for Elements constructed via `new` and not attached to another doc, which is a supported use case. And in that case we don't know either way.

I plan on making this namespace inference a defaulted-on option in the W3C DOM. This will allow existing uses who have working code without a namespace (for e.g. simple xpath queries) a migration path, or for new uses that prefer it otherwise.